File: api/routers/pagos.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_db
from dependencies.auth import get_current_user
from models.usuario import Usuario
from schemas.pago import DeudaPendienteResponse, MiPagoResponse
from schemas.suscripcion import PreviewRenovacionResponse, RenovacionSuscripcionPendienteResponse
from services.pago_service import PagoService
from services.suscripcion_service import SuscripcionService

router = APIRouter(prefix="/pagos", tags=["pagos"])


@router.get("/mios", response_model=list[MiPagoResponse])
async def get_mis_pagos(
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).get_mis_pagos(current_user)


@router.get("/renovaciones-pendientes", response_model=list[RenovacionSuscripcionPendienteResponse])
async def get_renovaciones_pendientes(
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await SuscripcionService(db).list_renovaciones_pendientes(current_user)


@router.get("/deudas-pendientes", response_model=list[DeudaPendienteResponse])
async def get_deudas_pendientes(
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).list_deudas_pendientes(current_user)


@router.post("/mp/preferencia")
async def crear_preferencia_mp(
    clase_template_id: UUID = Query(...),
    fecha: date = Query(...),
    monto: float = Query(..., gt=0),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).crear_preferencia_mp(current_user, clase_template_id, fecha, monto)


@router.post("/mp/confirmar")
async def confirmar_pago_mp(
    payment_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).confirmar_pago_mp(current_user, payment_id)


@router.post("/mp/preferencia-waitlist")
async def crear_preferencia_waitlist_mp(
    waitlist_id: UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).crear_preferencia_waitlist_mp(current_user, waitlist_id)


@router.post("/mp/confirmar-waitlist")
async def confirmar_waitlist_mp(
    payment_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).confirmar_waitlist_mp(current_user, payment_id)


@router.post("/mp/preferencia-deuda")
async def crear_preferencia_deuda_mp(
    asistencia_id: UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).crear_preferencia_deuda_mp(current_user, asistencia_id)


@router.post("/mp/confirmar-deuda")
async def confirmar_deuda_mp(
    payment_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).confirmar_deuda_mp(current_user, payment_id)


@router.post("/mp/preferencia-suscripcion")
async def crear_preferencia_suscripcion_mp(
    clase_template_id: UUID = Query(...),
    monto: float = Query(..., gt=0),
    fecha_inicio: date = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).crear_preferencia_suscripcion_mp(current_user, clase_template_id, monto, fecha_inicio)


@router.post("/mp/confirmar-suscripcion")
async def confirmar_suscripcion_mp(
    payment_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).confirmar_suscripcion_mp(current_user, payment_id)


@router.get("/preview-renovacion-suscripcion", response_model=PreviewRenovacionResponse)
async def preview_renovacion_suscripcion(
    suscripcion_id: UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).preview_renovacion_suscripcion(current_user, suscripcion_id)


@router.post("/mp/preferencia-renovacion-suscripcion")
async def crear_preferencia_renovacion_suscripcion_mp(
    suscripcion_id: UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).crear_preferencia_renovacion_suscripcion_mp(current_user, suscripcion_id)


@router.post("/mp/confirmar-renovacion-suscripcion")
async def confirmar_renovacion_suscripcion_mp(
    payment_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
):
    return await PagoService(db).confirmar_renovacion_suscripcion_mp(current_user, payment_id)

File: frontend/src/actions/pagos.ts
"use server";

import { serverApi } from "@/lib/server-api";
import { DeudaPendiente, MiPago, PreviewRenovacion, RenovacionSuscripcionPendiente } from "@/types/api";

export type PreferenciaResult = {
    init_point?: string;
    preference_id?: string;
    error?: string;
};

export type ConfirmarPagoResult = {
    status?: string;
    already_processed?: boolean;
    error?: string;
};

export async function getMisPagos(): Promise<MiPago[]> {
    try {
        const res = await serverApi("/pagos/mios");
        if (!res.ok) return [];
        return res.json();
    } catch {
        return [];
    }
}

export async function getRenovacionesPendientes(): Promise<RenovacionSuscripcionPendiente[]> {
    try {
        const res = await serverApi("/pagos/renovaciones-pendientes");
        if (!res.ok) return [];
        return res.json();
    } catch {
        return [];
    }
}

export async function getDeudasPendientes(): Promise<DeudaPendiente[]> {
    try {
        const res = await serverApi("/pagos/deudas-pendientes");
        if (!res.ok) return [];
        return res.json();
    } catch {
        return [];
    }
}

export async function crearPreferenciaMP(
    claseTemplateId: string,
    fecha: string,
    monto: number,
): Promise<PreferenciaResult> {
    try {
        const res = await serverApi("/pagos/mp/preferencia", {
            method: "POST",
            params: {
                clase_template_id: claseTemplateId,
                fecha,
                monto: monto.toString(),
            },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al iniciar el pago con MercadoPago" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function crearPreferenciaSuscripcionMP(
    claseTemplateId: string,
    monto: number,
    fechaInicio: string,
): Promise<PreferenciaResult> {
    try {
        const res = await serverApi("/pagos/mp/preferencia-suscripcion", {
            method: "POST",
            params: { clase_template_id: claseTemplateId, monto: monto.toString(), fecha_inicio: fechaInicio },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al iniciar el pago de suscripción" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function previewRenovacionSuscripcion(
    suscripcionId: string,
): Promise<PreviewRenovacion | { error: string }> {
    try {
        const res = await serverApi("/pagos/preview-renovacion-suscripcion", {
            params: { suscripcion_id: suscripcionId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al obtener el precio" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function crearPreferenciaRenovacionSuscripcionMP(
    suscripcionId: string,
): Promise<PreferenciaResult> {
    try {
        const res = await serverApi("/pagos/mp/preferencia-renovacion-suscripcion", {
            method: "POST",
            params: { suscripcion_id: suscripcionId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al iniciar la renovación" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function crearPreferenciaDeudaMP(asistenciaId: string): Promise<PreferenciaResult> {
    try {
        const res = await serverApi("/pagos/mp/preferencia-deuda", {
            method: "POST",
            params: { asistencia_id: asistenciaId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al iniciar el pago pendiente" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function confirmarDeudaMP(paymentId: string): Promise<ConfirmarPagoResult> {
    try {
        const res = await serverApi("/pagos/mp/confirmar-deuda", {
            method: "POST",
            params: { payment_id: paymentId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al confirmar el pago pendiente" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function confirmarSuscripcionMP(paymentId: string): Promise<ConfirmarPagoResult> {
    try {
        const res = await serverApi("/pagos/mp/confirmar-suscripcion", {
            method: "POST",
            params: { payment_id: paymentId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al confirmar la suscripción" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function confirmarRenovacionSuscripcionMP(paymentId: string): Promise<ConfirmarPagoResult> {
    try {
        const res = await serverApi("/pagos/mp/confirmar-renovacion-suscripcion", {
            method: "POST",
            params: { payment_id: paymentId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al confirmar la renovación" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}

export async function confirmarPagoMP(paymentId: string): Promise<ConfirmarPagoResult> {
    try {
        const res = await serverApi("/pagos/mp/confirmar", {
            method: "POST",
            params: { payment_id: paymentId },
        });
        if (!res.ok) {
            const error = await res.json().catch(() => ({}));
            return { error: error.detail || "Error al confirmar el pago" };
        }
        return res.json();
    } catch {
        return { error: "Error al conectar con el servidor" };
    }
}
